Source/NSWindow.m

```diff
@@ -4170,16 +4170,27 @@ window that has been ordered out (and thus is logically invisible). We
           switch (sub)
             {
             case GSAppKitWindowMoved:
-              _frame.origin.x = (CGFloat)[theEvent data1];
-              _frame.origin.y = (CGFloat)[theEvent data2];
-              NSDebugLLog(@"Moving", @"Move event: %d %@",
-                          (int)_windowNum, NSStringFromPoint(_frame.origin));
-              if (_autosaveName != nil)
-                {
-                  [self saveFrameUsingName: _autosaveName];
-                }
-              [nc postNotificationName: NSWindowDidMoveNotification
-                  object: self];
+              {
+                NSScreen *oldScreen;
+                NSScreen *newScreen;
+                oldScreen = [self screen];
```

Comment:
You could use the instance variable here. The frame may already have changed if the move was initiated by code.

Reply:
Indeed this is a problem. If position was changed programmatically (with setFrame:display:)  instance variable already changed. Probably I need to move this code into setFrame...

Reply:
Done. I've removed reassign of `_screen` in `setFrame:display:` and `setFrameFromString:` methods. Everything works correctly.
The results: 1. Less places in code where `_screen` is changing. 2. No need to use extra instance variable.

The idea behind this change is the following: change to window frame can be applied with these methods. Both methods will call to [GSDisplayServer placewindow::] and placewindow:: will generate GSAppKitWindowMoved event from -back to -gui. What do you think?

Reply:
Further thinking about it I've came to a conclusion that some logic should be quite different. For the case when window gets its frame programmatically window should be opened in screen where main menu resides (mainScreen). In that case origin of new window should adjusted to the frame of main screen.

It sounds good from usability perspective - user expects new window to appear on screen where main menu resides. 

I'm going to implement it and see how it works.

Reply:
There are different cases where the frame gets set programatically. The method setFrameFromString: is just one of them, center would be another.
And please also remember that there are other backends (MS Windows and soon Wayland), where the interaction between gui and back might be different.

Reply:
Yes, but center calls setFrame:display: finally. The call sequence is following `center -> setFrameOrigin: -> setFrame:display:`. That's why I think `setFrame:display:` is a good single place to tweak `_screen` value.

As for other backends: MS Windows should use in-window menus so this is not a case. Wayland - as new backend - should adopt current approach or propose new, better.

Reply:
OK, so back to the start, why don't you use the instance variable to get the oldScreen?

Reply:
No problem, done.